File: solutions/1.two-sum/hash-table.ts
/*
 * @lc app=leetcode.cn id=1 lang=typescript
 *
 * https://leetcode-cn.com/problems/two-sum
 */

// @lc code=start
function twoSum(nums: number[], target: number): number[] {
  const map: Map<number, number> = new Map();
  for (const [i, num] of nums.entries()) {
    const diff = target - num;
    if (map.has(diff)) {
      return [map.get(diff), i];
    }
    map.set(num, i);
  }
  return [];
}
// @lc code=end

File: solutions/1.two-sum/src/hash_table.rs
/*
 * @lc app=leetcode.cn id=1 lang=rust
 *
 * https://leetcode.cn/problems/two-sum/
 */

pub struct Solution;

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut map = HashMap::new();
        for (i, num) in nums.iter().enumerate() {
            let diff = target - num;
            if let Some(&j) = map.get(&diff) {
                return vec![j as i32, i as i32];
            }
            map.insert(num, i);
        }
        vec![]
    }
}
// @lc code=end
